R/prep_data_utils.R

```diff
@@ -64,3 +64,79 @@ find_new_missingness <- function(d, recipe) {
   new_missing <- dplyr::setdiff(missing_now, missing_then)
   return(dplyr::intersect(new_missing, predictors))
 }
+
+#' @title Convert character date columns to dates
+#'
+#' @description I try to convert columns ending in "DTS" to type Date. I'll take
+#' my best guess at the format and return a more standard one if I can. Times
+#' will be removed.
+#'
+#' @param d A dataframe or tibble containing data to try to convert to dates.
+#'
+#' @return A tibble containing the converted date columns. If no columns
+#' needed conversion, the original data will be returned.
+#' @import purrr
+#' @importFrom lubridate guess_formats date
+#' @export
+#'
+#' @examples
+#' d <- tibble::tibble(a_DTS = c("2018-3-25", "2018-3-25"),
+#' b_nums = c(2, 4),
+#' c_DTS = c("03-25-2018", "03-25-2018"),
+#' d_chars = c("a", "b"),
+#' e_date = lubridate::mdy(c("3-25-2018", "3-25-2018")))
+#'
+#' d <- convert_date_cols(d)
+#'
+convert_date_cols <- function(d) {
+
+  # Extract character date columns only
+  dd <- d[1, find_date_cols(d)]
+  col_names <- names(dd[!(map_lgl(dd, is.Date))])
+  dd <- dd[1, col_names]
+
+  date_formats <- map(dd,
+                      guess_formats,
+                      orders = c("ymd", "mdy", "ymd HMS", "mdy HMS"))
+
+  # Find working formats
+  valid_formats <- map2(dd, date_formats, function(x, y) {
+    temp <- map2(x, y, function(x, y) {
+      as.POSIXct(x = x, format = y)
+    })
+    # Can't use is.POSIXct here because NA has class POSIXct. wtf.
+    unlist(map(temp, function(x) {
+      !is.na(x)
+    }))
+  })
+
+  # Collapse to one format and find columns with no valid format
+  valid_formats <- map_int(valid_formats, function(x) {
+    x <- match(TRUE, x)
+  })
+  bad_date_cols <- names(valid_formats[is.na(valid_formats)])
+
+  if (length(bad_date_cols)) {
+    stop(paste("Unable to convert the following columns to dates. Convert",
+         "them to ymd, mdy, ymd hms, or mdy hms format. See lubridate::as_date",
+         "or as.POSIXct for more info. \n"), paste(bad_date_cols, collapse = ", "))
+  }
+
+  # Remove formats that don't work
+  date_formats <- date_formats[!is.na(valid_formats)]
```

Comment:
As above, date_formats and valid_formats have different length here. How about keeping everything in a data frame so that formats are organized by row and using `map(seq_len(nrow(d)), function(i) ...` when you need multiple aspects of the format at the same time?